dbuf-lsp/src/core/navigator/find_symbols.rs

```diff
@@ -0,0 +1,146 @@
+//! find all locations of symbol
+//! TODO:
+//! * maskings issues in patterns
+//!
+
+use tower_lsp::lsp_types::Range;
+
+use crate::core::ast_access::{LocStringHelper, LocationHelpers, Str};
+
+use crate::core::ast_visitor::scope_visitor::ScopeVisitor;
+use crate::core::ast_visitor::*;
+
+use super::Navigator;
+use super::Symbol;
+
+struct FindImpl<'a> {
+    target: &'a Symbol,
+    scope: ScopeVisitor<'a>,
+    ans: Vec<Range>,
+}
+
+pub fn find_symbols_impl(navigator: &Navigator, symbol: &Symbol) -> Vec<Range> {
+    let mut implementation = FindImpl {
+        target: symbol,
+        scope: ScopeVisitor::new(navigator.elaborated),
+        ans: Vec::new(),
+    };
+
+    visit_ast(navigator.parsed, &mut implementation, navigator.elaborated);
+
+    implementation.ans
+}
+
+impl FindImpl<'_> {
+    fn check_add(&mut self, str: &Str) {
+        match &self.target {
```

Comment:
Looks like a predicate which is hidden behind these `if-returns`. Maybe specify this predicate function and do `self.ans.push` if this predicate is (not) satisfied?